cosmos/x/evm/plugins/precompile/plugin.go

```diff
@@ -117,44 +118,66 @@ func (p *plugin) SetTransientKVGasConfig(transientKVGasConfig storetypes.GasConf
 // error or insufficient gas is provided.
 //
 // Run implements core.PrecompilePlugin.
+//
+//nolint:nonamedreturns // panic recovery.
 func (p *plugin) Run(
 	evm ethprecompile.EVM, pc vm.PrecompileContainer, input []byte,
-	caller common.Address, value *big.Int, suppliedGas uint64, readonly bool,
-) ([]byte, uint64, error) {
-	// use a precompile-specific gas meter for dynamic consumption
-	gm := storetypes.NewInfiniteGasMeter()
-	// consume static gas from RequiredGas
-	gm.ConsumeGas(pc.RequiredGas(input), "RequiredGas")
-
-	// get native Cosmos SDK context from the Polaris StateDB
+	caller common.Address, value *big.Int, suppliedGas uint64, readOnly bool,
+) (ret []byte, gasRemaining uint64, err error) {
+	// get native Cosmos SDK context and MultiStore from the Polaris StateDB
 	sdb := utils.MustGetAs[vm.PolarisStateDB](evm.GetStateDB())
 	ctx := sdk.UnwrapSDKContext(sdb.GetContext())
+	ms := utils.MustGetAs[MultiStore](ctx.MultiStore())
+
+	// make sure the readOnly is only set if we aren't in readOnly yet, which also makes sure that
+	// the readOnly flag isn't removed for child calls (taken from geth core/vm/interepreter.go)
+	if readOnly && !ms.IsReadOnly() {
+		ms.SetReadOnly(true)
+		defer func() { ms.SetReadOnly(false) }()
+	}
 
-	// disable reentrancy into the EVM
+	// disable reentrancy into the EVM only during precompile execution
 	p.disableReentrancy(sdb)
+	defer p.enableReentrancy(sdb)
+
+	// recover from any panic during precompile execution for the EVM to handle as a vm error
+	defer func() {
```

Comment:
I think we should abstract this into a panic handler, and then pass pointers in

Reply:
Fair